### PR TITLE
Mainmenu: Add customisable pager, add new change game dialog

### PR DESCRIPTION
**Pull request by rubenwardy**
This is the first PR in a series to improve the main menu.

This should be merged as two commits. 

The first commit splits up the `tabview` in such a way as to allow `layouters`. A layouter is an object with methods which are called to generate the formspec and handle events. The aim of this is to allow easy customisation of the navigation of the main menu. In future, games will be able to provide a Lua script at `menu/init.lua` to customise their main menu.

The second commit adds a new Change Game dialog, akin to celeron55's plans. This dialog will be the first thing that a *new* user sees when they start minetest, and will promote the use of ContentDB for installing games and such. Once a user has selected a game, future boots will load directly into that game. To get back to the change game dialog, they can select the layout's change game option.

image

image

## To do

- [x] Add tab view layouters
- [x] Add change game dialog
- [x] Fix game selection only working on local tab
- [x] Change selected game setting name so that users with existing installs will see the dialog
- [ ] Show change game dialog when `menu_current_game` is nil
- [ ] Implement "play online", "content", and "settings"

## How to test

WIP PR, testing later

### CODE REVIEW

**Comment by rubenwardy** [comment]
# Current Builtin Layouters

## tabs

The current system

## vertical

tabs are on the side, the original form of this PR. With formspec scaling, this would become the fullscreen version

image

image

## mainmenu

image

image

**Comment by Wuzzy2** [comment]
How do you select singleplayer games, though?

**Comment by ClobberXD** [comment]
> How do you select singleplayer games, though?

As far as I can tell, "New Game" to create the world, and "Load Game" to load an existing world.

**Comment by Wuzzy2** [comment]
That's not what I meant … I mean, how do you select a different game? How do you go from Minetest Game to MineClone 2?

**Comment by LoneWolfHT** [comment]
> That's not what I meant … I mean, how do you select a different game? How do you go from Minetest Game to MineClone 2?

I think what they said can still be your answer, just with different wording

**Comment by rubenwardy** [comment]
You click "Change Game" which then shows this screen:



This is WIP, the layout and the contrast isn't as good as it could be.

**Comment by Wuzzy2** [comment]
This is a much better layouting than our current one.

Random thoughts:
* The “You can download new games …” message: I'm not sure if it's neccessary. Sure, for new users it might be useful but after that it becomes a distraction. Maybe it could be a message that only appears once?
* Overall, all the menus need to become much more visually appalling, but I understand this is out of scope here. But I think it's important to keep in mind
* Change game dialog: Names need to be readable in full. Also, I'm kind of missing a game description here, but it would be very appropriate for such a menu to show the game description. I'm not sure about the layout. Maybe it's better to show games in a scrollable list (w/ icon, title and description).
* Main menu: Yes, 99% approval! This is the kind of main menu you know from most other games, so meets player's expectations. The only thing I'm missing is the “Quit” button. :-)
* A huge problem in UI design we have IMO is that we only have one font size for everything. A title has the same font size as everything else. This makes it difficult to implement e.g. captions and menu titles. E.g. the caption for the game selection dialog is very small

I have actually cloned your repo just to test this stuff myself, but it does not look at all like in the screenshots. :-(

**Comment by rubenwardy** [comment]
> Change game dialog: Names need to be readable in full. Also, I'm kind of missing a game description here, but it would be very appropriate for such a menu to show the game description. I'm not sure about the layout. Maybe it's better to show games in a scrollable list (w/ icon, title and description).

I'm planning to make it a vertical scrollable list with icon, title, and description - but I'm blocked by the scroll_container PR.

> I have actually cloned your repo just to test this stuff myself, but it does not look at all like in the screenshots. :-(

You need to fiddle with the code in `mainmenu/init.lua` to see the different styles, and the actual redesign part of it (ie: the content) is not part of this PR, it will be in a later PR

**Comment by v-rob** [comment]
> * Overall, all the menus need to become much more visually appalling

More _appalling_, or _appealing_? :)

> * A huge problem in UI design we have IMO is that we only have one font size for everything. A title has the same font size as everything else. This makes it difficult to implement e.g. captions and menu titles. E.g. the caption for the game selection dialog is very small

Setting font sizes is quite possible, as seen through the `hypertext` element, it's just a matter of adding that functionality to `label`.  Maybe through `style`.

**Comment by LoneWolfHT** [comment]
Maybe something like this, where all games are listed with their icons in that red bar, including the ones that haven't been downloaded (the icons for those will be greyscale'd to indicate that maybe)
image
Then down below in the blue area will be the game info, larger icon, name, short description, ect

**Comment by Wuzzy2** [comment]
No, definitely do NOT replicate the old “game button bar” like in our current main menu. It takes 1,000,000 clicks everytime to click through the game I want, it's very annoying.

**Comment by LoneWolfHT** [comment]
> No, definitely do NOT replicate the old “game button bar” like in our current main menu. It takes 1,000,000 clicks everytime to click through the game I want, it's very annoying.

It should just take one click for the scrollbar I didn't show at the top of the red bar and another click to click on the game icon?

**Comment by v-rob** [comment]
Yes, the button bar isn't that annoying.  I think using `scroll_container` would make it perfectly fine.  I personally don't like the idea of having to click on a button and then click on a game to get which one I want.  Not everyone has a ton of games; I never have more than three at a time, so a scrollable button bar is only one click for people like me and two (or a scroll and a click) for everyone else.  Clicking to get a new screen disrupts the mind, so it that will take even longer.

**Comment by rubenwardy** [comment]
image

**Comment by LoneWolfHT** [comment]
What would it look like if there were more games than space on the screen for them?

**Comment by rubenwardy** [comment]
It will scroll vertically using a scroll bar. This pr depends on scroll containers

**Comment by LoneWolfHT** [comment]
Might look better if there was a bit of a background in that case

**Comment by Wuzzy2** [comment]
> It will scroll vertically using a scroll bar. This pr depends on scroll containers

In the meantime, you could use pages instead.

**Comment by Wuzzy2** [comment]
One question:
What's the point of the left bar (“New Game”, “Load Game”, etc.) if you also did the good old main menu? Why have both and not one?

I would strongly prefer to use the classic main menu only. The tabs, or side buttons are pretty unique to Minetest, which is BAD because it will be a surprise to users.

You might argue that the point is to give games full freedom of designing their main menu but that could be actually a bad thing. Because then all games will just implement their own UI with their own, inconsistent conventions and the player must re-learn the UI from scratch every time.

**Comment by Wuzzy2** [comment]
@rubenwardy: About the game selection: Maybe only put 1 game per line? That way, there's much more room for the description. I would also make it more obvious that the game buttons are clickable, by making them look like a real clickable button and not just an icon.

**Comment by rubenwardy** [comment]
Im leaning towards making the main menu mode the default, as with some attention it would look the best.

I'll be looking into using a 9slice background for the buttons. Not sure how to combine that with the background

**Comment by rubenwardy** [comment]
image

**Comment by rubenwardy** [comment]
image

**Comment by rubenwardy** [comment]
image

**Comment by Wuzzy2** [comment]
I can't wait until this is in a testable state.

**Comment by Wuzzy2** [comment]
Oh, I noticed you removed the game name from the top left corner. Please bring it back.

The button “Play Online” is inconsitent with “Join Game”. In general, I'm not very happy with some of the menu names …

* Start Game: Obvious
* Join Game: Doesn't say it's online
* Content: Content of what? Way too vague.
* Settings: Obvious
* Change Game: Obvious
* Credits: Obvious

Also: Still no exit button. :P

**Comment by MoNTE48** [comment]
IMHO the interface in the first screenshot was simpler. I hate the back button. Back to the empty menu.
Switching between tabs now has 2 actions instead of one.

**Comment by cHyper-0815OL** [comment]
Are there any alpha versions of the new menu design out there for testing the new layout?

File: src/gui/guiBox.cpp [hunk 1]
```diff
@@ -31,6 +32,8 @@ void GUIBox::draw()
 	if (!IsVisible)
 		return;
 
+	// errorstream << "color: " << std::hex << m_color.color << std::endl;
```
Comment (appgurueu) [inline src/gui/guiBox.cpp:35]: Throw this and the `#include "log.h"` out?

**Comment by Pevernow** [comment]
It seems all right.
But I don't know why it was close?

**Comment by sfan5** [comment]
I believe rubenwardy currently has no time to work on this.

**Comment by Tim79000** [comment]
I hope this gets added, but a few gripes, 1, will their be an option for showing games in a list without icons?
2, how in depth would game menu customization be?

**Pull request closed without merging**